### PR TITLE
feat(papi): add labware.set_offset for jupyter/cli, bump papi to 2.12

### LINKED ISSUE #9171: Consume Labware Offsets in Jupyter
Create a method to consume labware offsets (downloaded from the runapp as defined in #9055) in Jupyter notebook.

### PR DESCRIPTION
**Pull request by mcous**
## Overview

In order to support using LPC offsets in Jupyter and CLI workflows, this PR adds a new public method to the `Labware` PAPI class.

```py
@requires_version(2, 12)
def set_offset(self, x: float, y: float, z: float) -> None:
    ...
```

It's based on the private, now unused `Labware.set_calibration` method. In contrast:

- This method is public
- It does not require importing the `Point` class
- It is named `set_offset` for consistency with LPC language

See https://github.com/Opentrons/opentrons/issues/9059#issuecomment-1011286300 for example usage in Jupyter and CLI flows.

Closes #9171, #9172

## Changelog

- feat(papi): add labware.set_offset for jupyter/cli, bump papi to 2.12

## Review requests

- Does this line up with our previous discussions?
- Tests / docs look good

## Risk assessment

Low

### CODE REVIEW

**Comment by sanni-t** [comment]
Tested on a robot using the following steps (Jupyter) / protocol (via `opentrons_execute`)
```
from opentrons import protocol_api

metadata = {
	"apiLevel": "2.12"
}

def run(protocol: protocol_api.ProtocolContext):
	labware1 = protocol.load_labware('opentrons_96_tiprack_10ul', 2)
	p20 = protocol.load_instrument('p20_single_gen2', 'right', tip_racks=[labware1])
	protocol.home()
	p20.move_to(labware1.wells()[0].top())
	p20.home()
	labware1.set_offset(x=2.0, y=-3.0, z=10.0)
	p20.move_to(labware1.wells()[0].top())
```
Pipette correctly moved to a point offset by the amount set (eyeballed) on the second `move_to` command on both Jupyter and ssh execution.

**Comment by sanni-t** [comment]
Also tested the same protocol except with `labware1` loaded on a temperature module and the offset still applied.

File: api/src/opentrons/protocol_api/labware.py [hunk 1]
```diff
@@ -337,6 +337,16 @@ def set_calibration(self, delta: Point):
...
         self._implementation.set_calibration(delta)
 
+    @requires_version(2, 12)
+    def set_offset(self, x: float, y: float, z: float) -> None:
+        """Set the labware's position offset directly.
+
+        This method is primarily for Jupyter and command-line applications
+        of the Python Protocol API. If you're not using Jupyter or the CLI,
+        use Labware Position Check in the Opentrons App, instead.
+        """
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/labware.py]: Some suggestions:

-----

Nitpick: I'm not sure the word "directly" is doing much here; I'd just drop it maybe?

-----

Can we clarify that `x`, `y`, and `z` are:

* in deck coordinates (you can cross-ref to those docs like ``see :ref:`protocol-api-deck-coords` ``
* relative to the labware's normal position within its deck slot or module, as opposed to being global


-----

Can we specify what happens if you use this in a protocol uploaded through the Opentrons App?

Options:

* Specify that it *overwrites* the offset supplied by Labware Position Check. (I believe this is how this implementation will currently behave.)
* Specify that it *adds to* the offset supplied by Labware Position Check.
* Specify that you can't do it, and make the code raise an error if you try.
* Specify that you can't do it, but don't enforce that in the code. i.e., make it undefined behavior.

**Review by SyntaxColoring — APPROVED** [review]
API design and tests look good to me. Some suggestions for the docstring.

I also think the "tutorial-style" part of our docs should link to this reference documentation, as part of the user's general introduction to using Jupyter. Unfortunately, our Jupyter docs are scattered across various places on docs.opentrons.com and support.opentrons.com ([example 1](https://docs.opentrons.com/v2/new_advanced_running.html#jupyter-notebook), [example 2](https://support.opentrons.com/en/articles/3136506-using-labware-in-your-protocols), [example 3](https://support.opentrons.com/en/articles/1795367-running-the-robot-using-jupyter-notebook#h_216e65fa4e)), so to avoid making things worse, we'll want to coordinate with Support before doing that. (None of that has to happen in this PR, to be clear.)

File: api/src/opentrons/protocol_api/labware.py [hunk 2]
```diff
@@ -337,6 +337,28 @@ def set_calibration(self, delta: Point):
...
+        even if those labware are of the same type.
+
+        .. caution::
+            This method is *only* for Jupyter and command-line applications
+            of the Python Protocol API. Do not use this method in a protocol
+            uploaded via the Opentrons App.
+
+            Using this method and the Opentrons App's Labware Position Check
+            at the same time will produce undefined behavior. We may choose
+            to define this behavior in a future release
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/labware.py]:
```suggestion
            Using this method and the Opentrons App's Labware Position Check
            at the same time will produce undefined behavior. We may choose
            to define this behavior in a future release.
```

**Review by SyntaxColoring — APPROVED** [review]
Changes look good to me. Thanks!

In the rendered docs, the "added in version..." thing is getting inserted into the `caution` block for some reason, instead of coming after it. The `@requires_version` source looks like it *intends* to account for this edge case, so I don't know why that isn't working. 🤷

**Pull request merged**